mm2src/mm2_db/src/indexed_db/drivers/cursor/cursor.rs

```diff
@@ -241,72 +242,91 @@ impl CursorDriver {
-            }
+            })?,
+            CursorAction::ContinueWithValue(next_value) => {
+                cursor
+                    .continue_with_key(next_value)
+                    .map_to_mm(|e| CursorError::AdvanceError {
+                        description: stringify_js_error(&e),
+                    })?
+            },
+            // Don't advance the cursor.
+            // Here we set the `stopped` flag so we return `Ok(None)` at the next iteration immediately.
+            // This is required because `item_action` can be `CollectItemAction::Include`,
+            // and at this iteration we will return `Ok(Some)`.
+            CursorAction::Stop => self.stopped = true,
+        }
+
+        Ok(())
+    }
+
+    async fn process_cursor_item(&mut self, where_: Option<&CursorCondition>) -> CursorResult<Option<(ItemId, Json)>> {
+        let event = match self.cursor_item_rx.next().await {
+            Some(event) => event,
+            None => {
+                self.stopped = true;
+                return Ok(None);
+            },
+        };
+
+        let _cursor_event = event.map_to_mm(|e| CursorError::ErrorOpeningCursor {
+            description: stringify_js_error(&e),
+        })?;
 
-            match item_action {
-                CursorItemAction::Include => return Ok(Some(item.into_pair())),
-                // Try to fetch the next item.
-                CursorItemAction::Skip => (),
+        let cursor = match cursor_from_request(&self.cursor_request)? {
+            Some(cursor) => cursor,
+            None => {
+                self.stopped = true;
+                return Ok(None);
+            },
+        };
+
+        let (key, js_value) = match (cursor.key(), cursor.value()) {
+            (Ok(key), Ok(js_value)) => (key, js_value),
+            _ => {
+                self.stopped = true;
+                return Ok(None);
+            },
+        };
+
+        let item: InternalItem =
+            deserialize_from_js(js_value).map_to_mm(|e| CursorError::ErrorDeserializingItem(e.to_string()))?;
+        let (item_action, cursor_action) = self.inner.on_iteration(key)?;
+
+        let (id, val) = item.into_pair();
+        // Checks if the given `where_` condition, represented by an optional closure (`cursor_condition`),
+        // is satisfied for the provided `item`. If the condition is met, return the corresponding `(id, val)` or skip to the next item.
+        if let Some(cursor_condition) = where_ {
+            if cursor_condition(val.clone())? && matches!(item_action, CursorItemAction::Include) {
+                return Ok(Some((id, val)));
             }
+            self.continue_(&cursor, &cursor_action).await?;
+            return Ok(None);
         }
+
+        self.continue_(&cursor, &cursor_action).await?;
+        let res = match item_action {
+            CursorItemAction::Include => Some((id, val)),
+            CursorItemAction::Skip => None,
+        };
+        Ok(res)
```

Comment:
We already check if `CursorItemAction` is include so no need for the `match item_action`if we do the below
```rust
if matches!(item_action, CursorItemAction::Include) {
    if let Some(cursor_condition) = where_ {
        if cursor_condition(val.clone())? {
            return Ok(Some((id, val)));
        }
    } else {
        return Ok(Some((id, val)));
    }
}
self.continue_(&cursor, &cursor_action).await?;
Ok(None)
```
Also checking if `CursorItemAction` is include first is cheaper than checking `cursor_condition` first, so this solves 2 problems :)

Reply:
we want to call `self.continue_` before result, since entering the loop again to get next data will same return same old data I think as we didn't advance the cursor yet
 
```rust
 return Ok(Some((id, val)));
```